Implement bootstrap msm

msultan opened the pull request:
This is the initial commit for adding a msmb class for creating families of estimators via bootstraping over sequence space. 
I think this is what the api should look like. 

```python
from msmbuilder.msm import MarkovStateModel
from msmbuilder.msm.bootstrapmodel import BootstrapModel
msm = MarkovStateModel
bmsm = BootstrapModel(estimator = msm, n_samples = 10)
bmsm.fit(sequences)
```

msultan commented:
Ideally, this class would contain methods for comparing the mle estimator to the bootstrapped models and a nice mapping scheme from input space to all model space.

cxhernandez commented:
Just organizationally, I feel like this should be it's own submodule under `msmbuilder`, e.g.:

`from msmbuilder.bootstrapmodel import BootstrapModel `

or

`from msmbuilder import BootstrapModel `

msultan commented:
alright, i have changed it a bit so that the ```BootStrapMarkovStateModel``` mdl fits both the mle MSM and the bootstraped mdls in one go and adds some convenience attributes for comparing populations across all mdls. 
```
python
from msmbuilder.msm.validation import BootStrapMarkovStateModel
%pylab inline
bsmsm = BootStrapMarkovStateModel(lag_time=10,verbose=False,n_procs=2)
bsmsm.fit(seq)
```

I also added some 
errorbar(range(bsmsm.mle.n_states_),bsmsm.mapped_populations_mean_, 
         bsmsm.mapped_populations_sem_)

msultan commented:
okay i think this is ready for code review.

mpharrigan commented:
remove wip :)

cxhernandez commented on msmbuilder/msm/validation/bootstrapmsm.py:
```diff
@@ -0,0 +1,135 @@
+        self.mapped_populations_ = None
+        self.resample_ind_ = None
+
+    def fit(self, sequences, y=None):
```
Maybe there should be a `parallel` flag, which is default `True`?

msultan replied:
Wouldn't n_procs=1 serve the same purpose? I am tempted to default to multiprocessing so that this fits in a somewhat reasonable time for larger datasets.

cxhernandez replied:
> Wouldn't n_procs=1 serve the same purpose? 

Ah, that's true. Nevermind then.

> I am tempted to default to multiprocessing

Go for it :smile:

msultan commented:
fantastic, py2.7 cant pickle the _fit_one method. Gonna try and fix this later in the day.

mpharrigan commented on msmbuilder/msm/validation/__init__.py:
```diff
@@ -0,0 +1,2 @@
+from __future__ import absolute_import
+from .bootstrapmsm import *
```
Don't import `*`
(this pulls in Pool, cpu_count, list_of_1d, MarkovStateModel et al. as well)
(you could define `__all__` in bootstrapmsm.py but why not just import the one class explicitly)

mpharrigan commented on msmbuilder/msm/validation/bootstrapmsm.py:
```diff
@@ -0,0 +1,164 @@
+            mdl.fit(sequences)
+            # solve the eigensystem
+            mdl.eigenvalues_[0]
+        except:
```
No bare `try...except`. This might hide real issues! Be specific about what exception you're expecting. If it makes sense, add a warning as well

msultan replied:
ohh good point @mpharrigan . i was thinking of the value error that pops up when fitting msms every now and then. Gonna change it to explicitly expect the value error. .

mpharrigan commented on msmbuilder/tests/test_bootstrap_msm.py:
```diff
@@ -0,0 +1,90 @@
+    eq(mle_mdl.timescales_, b_mdl.mle_.timescales_)
+
+
+class fakemsm():
```
inherit from `object` for py2

mpharrigan commented on msmbuilder/tests/test_bootstrap_msm.py:
```diff
@@ -0,0 +1,90 @@
+    assert (mapped_pop==[0.1, 0, 0.3]).all()
+
+def test_mapper_4():
+   #case where the mdl is jumbled around
```
indent this to the same level

mpharrigan commented on msmbuilder/msm/validation/bootstrapmsm.py:
```diff
@@ -0,0 +1,168 @@
+
+
+def _fit_one(jt):
+    print(jt)
```
is there a reason why you're printing here?

msultan replied:
shit, forgot about that. Was debugging on py27. Thanks for the catch

mpharrigan commented:
:+1:

msultan commented:
Merged!
